aarch64: relax cache topology restrictions

dianpopa opened the pull request:
# Reason for This PR

Fixes #2407 

## Description of Changes

* the commits contain the info needed

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [ ] All commits in this PR are signed (`git commit -s`).
- [ ] The reason for this PR is clearly provided (issue no. or explanation).
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] Any newly added `unsafe` code is properly documented.
- [ ] Any API changes are reflected in `firecracker/swagger.yaml`.
- [ ] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [ ] All added/changed functionality is tested.

georgepisaltu commented on src/arch/src/aarch64/fdt.rs:
```diff
@@ -338,17 +338,27 @@ fn create_cpu_nodes(fdt: &mut Vec<u8>, vcpu_mpidr: &[u64]) -> Result<()> {
-                cache.type_.of_cache_sets(),
-                cache.number_of_sets as u32,
-            )?;
+            if cache.size_.is_some() {
+                append_property_u32(
+                    fdt,
+                    cache.type_.of_cache_size(),
+                    cache.size_.unwrap() as u32,
+                )?;
+            }
```
nit: As this is used multiple times below (and in similar fashion), maybe this can be a function or macro for deduplication and to avoid the use of `unwrap` with an `if let`.

dianpopa commented:
@georgepisaltu addressed your feedback and [pclesr's feedback](https://github.com/firecracker-microvm/firecracker/issues/2407#issuecomment-777470533). PTAL!

alindima commented on src/arch/src/aarch64/cache_info.rs:
```diff
@@ -101,9 +118,76 @@ impl CacheType {
+            *err_str += "number of sets";
+        }
+
+        Ok(())
```
I find it a little weird to use this `err_str` and return an `Ok` value. Is there a reason we're not using a regular `Error`?

alindima replied:
I understand what you're saying. However, I still find it confusing to have an `err_str` with an `Ok()` value.
Can't we distinguish the error types in the caller of this function? We can then match on the returned Error and decide whether it's a soft error (one that does not kill the process) or a hard error.

alindima replied:
looks good

alindima commented on src/arch/src/aarch64/cache_info.rs:
```diff
@@ -101,9 +118,76 @@ impl CacheType {
+                Error::InvalidCacheAttr("coherency_line_size".to_string(), e.to_string())
+            })?);
+        } else {
+            *err_str += if *err_str != "" { ", " } else { "" };
```
we can simplify this logic by simply adding a comma after each append operation, and pop the last comma right before returning

alindima commented on src/arch/src/aarch64/cache_info.rs:
```diff
@@ -318,24 +356,208 @@ mod tests {
+        let mut non_l1_caches: Vec<CacheInfo> = Vec::new();
+        // We use sysfs for extracting the cache information.
+        assert!(sysfs_read_caches(&mut l1_caches, &mut non_l1_caches).is_ok());
+        assert_eq!(l1_caches.len(), 2);
```
Are these assumptions we're willing to make? These depend on the cache topology of the host, right?

If somebody is running these unit tests on another machine that does not have these, they will run into errors I'm assuming.

Can we somehow mock the sysfs for this test?

dianpopa replied:
Indeed, these asserts depend on the host configuration and they might fail when run on other machines. I removed them.
Mocking a sysfs is not quite obvious given that `sysfs_read_caches` (the function that we are testing) constructs the `cache` object inside it and so I do not have access to the `cache_dir` field.

alindima replied:
I see. However, can't we mock the `cache_dir` by implementing the `Default` trait in the test mod with a custom path?

dianpopa reviewed (commented):
@alindima @georgepisaltu I addressed your latest feedback, please give it another look. Thanks

dianpopa reviewed (commented):
Regarding https://github.com/firecracker-microvm/firecracker/pull/2438#discussion_r587413612, re-implementing `Default` in the tests has a tricky part associated. `TempDir` will [remove the associated path once it goes out of scope](https://github.com/rust-vmm/vmm-sys-util/blob/master/src/unix/tempdir.rs#L26). So, when `sysfs_read_caches` will try to make use of `cache_dir` will crash. Unfortunately, the `vmm-sys-util` implementation of `TempDir` does not have support for making the path persist after drop. 
Now, that I think about the big picture, this function is also called from the fdt code which will also obtain an FDT specific to the machine that our CI is run on (i.e m6g). So, my proposal here is to tackle this separately as part of this [issue](https://github.com/firecracker-microvm/firecracker/issues/2492) since it has a larger scope, WDYT?

alindima commented:
> Regarding [#2438 (comment)](https://github.com/firecracker-microvm/firecracker/pull/2438#discussion_r587413612), re-implementing `Default` in the tests has a tricky part associated. `TempDir` will [remove the associated path once it goes out of scope](https://github.com/rust-vmm/vmm-sys-util/blob/master/src/unix/tempdir.rs#L26). So, when `sysfs_read_caches` will try to make use of `cache_dir` will crash. Unfortunately, the `vmm-sys-util` implementation of `TempDir` does not have support for making the path persist after drop.
> Now, that I think about the big picture, this function is also called from the fdt code which will also obtain an FDT specific to the machine that our CI is run on (i.e m6g). So, my proposal here is to tackle this separately as part of this [issue](https://github.com/firecracker-microvm/firecracker/issues/2492) since it has a larger scope, WDYT?

I agree
